ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Linux/procinfo.cpp

```diff
@@ -0,0 +1,159 @@
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in all
+ copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#include <cstdlib>
+#include <cstring>
+
+#include "../procinfo.h"
+
+#include <proc/readproc.h>
+
+using std::string;
+using std::vector;
+using std::to_string;
+
+namespace procinfo {
+
+string path_from_pid(process_t pid) {
+  string path;
+  string link = string("/proc/") + to_string(pid) + string("/exe");
+  char *buffer = realpath(link.c_str(), NULL);
+  path = buffer ? : "";
+  free(buffer);
+  return path;
+}
+
+string cmd_from_pid(process_t pid) {
+  string cmd;
+  proc_t proc_info;
+  if (!pid_exists(pid)) { return ""; }
+  memset(&proc_info, 0, sizeof(proc_info));
+  PROCTAB *proc = openproc(PROC_FILLCOM | PROC_PID, &pid);
+  if (readproc(proc, &proc_info) != 0) {
+    char **args = proc_info.cmdline;
+    size_t cnta = 0;
+    for (; args[cnta] != NULL && strlen(args[cnta]) != 0; cnta++);
+    for (size_t i = 0; i < cnta; i++) {
+      if (string_has_whitespace(args[i])) {
+        cmd += "\"" + string_replace_all(args[i], "\"", "\\\"") + "\"";
+      } else {
+        cmd += args[i];
+      }
+      if (args[i + 1] != NULL) {
+        cmd += " ";
+      }
+    }
+  }
+  closeproc(proc);
+  return cmd;
+}
+
+string env_from_pid(process_t pid) {
+  string env;
+  proc_t proc_info;
+  if (!pid_exists(pid)) { return ""; }
+  memset(&proc_info, 0, sizeof(proc_info));
+  PROCTAB *proc = openproc(PROC_FILLENV | PROC_PID, &pid);
+  if (readproc(proc, &proc_info) != 0) {
+    char **envs = proc_info.environ;
+    size_t cnta = 0;
+    for (; envs[cnta] != NULL && strlen(envs[cnta]) != 0; cnta++);
+    for (size_t i = 0; i < cnta; i++) {
+      unsigned j = 0;
+      if (string_has_whitespace(envs[i]) || string_count_equalssigns(envs[i]) > 1) {
+        vector<string> envVec = string_split_by_first_equalssign(envs[i]);
+        for (const string &environ : envVec) {
+          if (j == 0) { env += environ; }
+          else { env += "=\"" + string_replace_all(environ, "\"", "\\\"") + "\""; }
+          j++;
+        }
+        j = 0;
+      } else {
+        env += envs[i];
+      }
+      if (envs[i + 1] != NULL) {
+        env += "\n";
+      }
+    }
+  }
+  closeproc(proc);
+  return env;
+}
+
+string pids_enum(bool trim_dir, bool trim_empty) {

```

Comment:
@RobertBColton I haven't touched the code used in this function in weeks, perhaps even more than a month. Any idea why since yesterday, calling pids_enum on Mac went from working yesterday and every day before that to always crashing, starting today? Am I using `new` and `delete` correctly?